aws/src/test/java/org/apache/iceberg/aws/TestS3FileIOSharedResourceManagement.java

```diff
@@ -0,0 +1,249 @@
+    S3FileIO originalFileIO = new S3FileIO();
+    originalFileIO.initialize(properties);
+
+    // Serialize the S3FileIO
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    try (ObjectOutputStream oos = new ObjectOutputStream(baos)) {
+      oos.writeObject(originalFileIO);
+    }
+
+    // Deserialize on "executor"
+    ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
+    S3FileIO deserializedFileIO;
+    try (ObjectInputStream ois = new ObjectInputStream(bais)) {
+      deserializedFileIO = (S3FileIO) ois.readObject();
+    }
+
+    // Both should work without conflicts
+    assertThat(originalFileIO.client()).isNotNull();
+    assertThat(deserializedFileIO.client()).isNotNull();
+
+    // Close original (simulating broadcast cleanup) - deserialized should still work
+    originalFileIO.close();
+    assertThatCode(deserializedFileIO::client).doesNotThrowAnyException();
+
+    deserializedFileIO.close();
+  }
+
+  @Test
+  public void testConcurrentAccessFromMultipleThreads() throws InterruptedException {
+    int threadCount = 10;
+    ExecutorService executor = Executors.newFixedThreadPool(threadCount);
+    CountDownLatch startLatch = new CountDownLatch(1);
+    CountDownLatch completionLatch = new CountDownLatch(threadCount);
+    AtomicInteger successCount = new AtomicInteger(0);
+    AtomicReference<Exception> firstException = new AtomicReference<>();
+
+    // Create multiple threads that each create and use S3FileIO instances
+    for (int i = 0; i < threadCount; i++) {
+      executor.submit(
+          () -> {
+            try {
+              startLatch.await(); // Wait for all threads to be ready
+
+              S3FileIO fileIO = new S3FileIO();
+              fileIO.initialize(properties);
+
+              // Simulate some operations
+              assertThat(fileIO.client()).isNotNull();
+
+              // Simulate random close timing (like Spark GC cleanup)
+              if (Math.random() > 0.5) {
+                Thread.sleep((long) (Math.random() * 100));
+              }
+
+              fileIO.close();
+              successCount.incrementAndGet();
+
+            } catch (Exception e) {
+              firstException.compareAndSet(null, e);
+            } finally {
+              completionLatch.countDown();
+            }
+          });
+    }
+
+    // Start all threads simultaneously
+    startLatch.countDown();
+
+    // Wait for completion
+    assertThat(completionLatch.await(30, TimeUnit.SECONDS)).isTrue();
+
+    // Verify no exceptions occurred
+    if (firstException.get() != null) {
+      throw new RuntimeException("Test failed with exception", firstException.get());
+    }
+
+    assertThat(successCount.get()).isEqualTo(threadCount);
+
+    executor.shutdown();
+    assertThat(executor.awaitTermination(5, TimeUnit.SECONDS)).isTrue();
+  }
+
+  @Test
+  public void testMemoryPressureSimulation() throws InterruptedException {
```

Comment:
I'm pretty hesitant to add tests like this though I understand the degree of confidence it builds. The issue is across different environments, this test may not be reproducible right? We're not really guaranteed that System.gc does anything etc.

Reply:
Agreed, I removed this in the new revision since I don't depend on Cleaner interface anymore.